### PR TITLE
[msbuild] Make sure to codesign appex dylibs

### PR DESCRIPTION
**Pull request by jstedfast**
Fixes https://bugzilla.xamarin.com/show_bug.cgi?id=52745

Added a new CodesignNativeLibraries task that scans for
and then codesigns each *.dylib and *.metallib in the
app bundle (minus those in the PlugIns and Watch dirs).

### CODE REVIEW

**Comment by jstedfast** [comment]
I'm also wondering if perhaps this new task should replace the logic that codesigns frameworks...?

Since we have to scan the bundle dir anyway...

**Comment by jstedfast** [comment]
/cc @kzu @joj 

Cc'ing you guys because this will require changes to the VS side of things, too.

Keep in mind that this patch will need to be back-ported to the 15.1 branch as well.

File: msbuild/tests/Xamarin.iOS.Tasks.Tests/ProjectsTests/CodesignAppBundle.cs [hunk 1]
```diff
@@ -1,105 +1,147 @@
...
 			// Rebuild w/ no changes
-			BuildProject ("MyTabbedApplication", Platform, "Release", clean: false);
+			BuildProject ("MyTabbedApplication", Platform, config, clean: false);
+
+			AssertProperlyCodesigned ();
 
-			var newTimestamps = Directory.EnumerateFiles (AppBundlePath, "*.*", SearchOption.AllDirectories).ToDictionary (file => file, file => GetLastModified (file));
-			var newDsymTimestamps = Directory.EnumerateFiles (dsymDir, "*.*", SearchOption.AllDirectories).ToDictionary (file => file, file => GetLastModified (file));
-			var newAppexDsymTimestamps = Directory.EnumerateFiles (appexDsymDir, "*.*", SearchOption.AllDirectories).ToDictionary (file => file, file => GetLastModified (file));
+			// Note: the following checks all require msbuild instead of xbuild to work properly
```
Comment (rolfbjarne) [inline msbuild/tests/Xamarin.iOS.Tasks.Tests/ProjectsTests/CodesignAppBundle.cs]: Can you check at runtime if MSBuild is used and only execute the asserts in that case?

I'm fairly sure we'll forget to uncomment when we switch to msbuild, so a runtime check would be better...

**Comment by kzu** [comment]
@emaf @josegallardo we need to adapt to this on our side before/when we bump.

**Comment by jstedfast** [comment]
@rolfbjarne You don't sign dylibs with a provisioning profile. You only need to sign them with the appropriate signing key. And yes, it will sign with the appropriate signing key as specified by the App Extension's project configuration settings (note the `AppExtensionCodesignProperties` values).

The reason that the task is skipping the `PlugIns` and `Watch` directories is because when you sign a bundle, you do not want to re-sign any watch or app extensions that are children of that bundle as it would invalidate their signatures.

Remember: this task gets called on each and every app extension as well as the main app bundle.

File: msbuild/tests/Xamarin.iOS.Tasks.Tests/ProjectsTests/CodesignAppBundle.cs (hunk 1 above)
Comment (jstedfast) [inline msbuild/tests/Xamarin.iOS.Tasks.Tests/ProjectsTests/CodesignAppBundle.cs]: Until we merge @radical's PR, I'm not sure how to do this... or if it's even possible.

**Pull request merged**